File: trunk/csharp/UrakawaToolkit/MetadataFactory.cs
using System;

namespace urakawa.project
{  
	/// <summary>
	/// Default <see cref="IMetadata"/> factory - supports creation of <see cref="Metadata"/> instances
	/// </summary>
	public class MetadataFactory : IMetadataFactory
	{
    /// <summary>
    /// Default constructor
    /// </summary>
		public MetadataFactory()
		{
    }
    #region IMetadataFactory Members

    /// <summary>
    /// Creates a <see cref="IMetadata"/> instance of the given <see cref="MetadataType"/> string representation
    /// </summary>
    /// <param name="type">The given metadata type</param>
    /// <returns>The created instance</returns>
    public IMetadata createMetadata(string type)
    {
      switch (type)
      {
        case "Metadata":
          return new Metadata();
        default:
          throw new exception.OperationNotValidException(
            String.Format("Can not create IMetadata instance of type {0}", type));
      }
    }

    #endregion
  }
}

File: java/UMLGraph/src/org/daisy/urakawa/metadata/MetadataFactory.java
package org.daisy.urakawa.metadata;

import org.daisy.urakawa.WithPresentation;
  
import org.daisy.urakawa.exception.MethodParameterIsEmptyStringException;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.xuk.XukAble;

/**
 * <p>
 * This is the factory that creates {@link org.daisy.urakawa.metadata.Metadata}
 * instances.
 * </p>
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Aggregation 1 org.daisy.urakawa.Project
 * @depend - Create - org.daisy.urakawa.metadata.Metadata
 */
public interface MetadataFactory extends XukAble, WithPresentation {
	/**
	 * <p>
	 * Creates a new metadata instance.
	 * </p>
	 * <p>
	 * This factory method does not take any argument and creates an object of
	 * the default type.
	 * </p>
	 * 
	 * @return cannot return null.
	 */
	public Metadata createMetadata();

	/**
	 * <p>
	 * Creates a new metadata instance.
	 * </p>
	 * 
	 * <p>
	 * This factory method takes arguments to specify the exact type of object
	 * to create, given by the unique QName (XML Qualified Name) used in the XUK
	 * serialization format. This method can be used to generate instances of
	 * subclasses of the base object type.
	 * </p>
	 * @param xukLocalName
	 *            cannot be null, cannot be empty string.
	 * @param xukNamespaceURI
	 *            cannot be null, but can be empty string.
	 * @return can return null (in case the QName specification does not match
	 *         any supported type).
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 * @throws MethodParameterIsEmptyStringException
	 *             Empty string '' method parameter is forbidden:
	 *             <b>xukLocalName</b>
	 * @tagvalue Exceptions "MethodParameterIsNull-MethodParameterIsEmptyString"
	 */
	public Metadata createMetadata(String xukLocalName, String xukNamespaceURI)
			throws MethodParameterIsNullException,
			MethodParameterIsEmptyStringException;
}
